config/configmapprovider/expand.go

```diff
@@ -17,52 +17,79 @@ package configmapprovider // import "go.opentelemetry.io/collector/config/config
 import (
 	"context"
 	"os"
+	"strings"
 
 	"go.opentelemetry.io/collector/config"
 )
 
-// NewExpandConverter returns a service.ConfigMapConverterFunc, that expands all environment variables for a given config.Map.
+// NewExpandConverter returns a config.MapConverterFunc, that expands all environment variables for a given config.Map.
+//
+// The `configMapProviders` is a map of pairs <scheme,Provider>.
 //
 // Notice: This API is experimental.
-func NewExpandConverter() config.MapConverterFunc {
-	return func(_ context.Context, cfgMap *config.Map) error {
+func NewExpandConverter(configMapProviders map[string]Provider) config.MapConverterFunc {
+	// TODO: Figure out how to support watching for changes in a Converter, or if expand should be a converter.
+	return func(ctx context.Context, cfgMap *config.Map) error {
 		for _, k := range cfgMap.AllKeys() {
-			cfgMap.Set(k, expandStringValues(cfgMap.Get(k)))
+			cfgMap.Set(k, expandStringValues(cfgMap.Get(k), func(location string) string {
+				// This allows escaping environment variable substitution via $$, e.g.
+				// - $FOO will be substituted with env var FOO
+				// - $$FOO will be replaced with $FOO
+				// - $$$FOO will be replaced with $ + substituted env var FOO
+				if location == "$" {
+					return "$"
+				}
+				scheme := "env"
+				if idx := strings.Index(location, ":"); idx != -1 {
+					scheme = location[:idx]
+				} else {
+					// MapProviders require the location to always have the scheme.
+					location = scheme + ":" + location
+				}
+				p, ok := configMapProviders[scheme]
+				// TODO: Figure out how to return an error when not registered.
+				if !ok {
+					return ""
+				}
+				// TODO: Figure out how to return an error when retrieve returns an error.
```

Comment:
probably would need to register an error callback somewhere along the way here.

Reply:
No longer needed.